pkg/flowaggregator/flowaggregator.go

```diff
@@ -15,25 +15,250 @@
+		collectingProtocol,
+		nil,
+		nil,
 		exportInterval,
+		nil,
+		0,
+		registry,
+	}
+	return fa
+}
+
+func genObservationID() (uint32, error) {
+	name, err := env.GetNodeName()
+	if err != nil {
+		return 0, err
+	}
+	h := fnv.New32()
+	h.Write([]byte(name))
+	return h.Sum32(), nil
+}
+
+func (fa *flowAggregator) InitCollectingProcess() error {
+	var collectAddr net.Addr
+	var err error
+	if fa.collectingProtocol == "tcp" {
+		collectAddr, _ = net.ResolveTCPAddr("tcp", "0.0.0.0:4739")
+		fa.collectingProcess, err = ipfix.NewIPFIXCollectingProcess(collectAddr, 65535, 0)
+	} else {
+		collectAddr, _ = net.ResolveUDPAddr("udp", "0.0.0.0:4739")
+		fa.collectingProcess, err = ipfix.NewIPFIXCollectingProcess(collectAddr, 1024, 0)
 	}
+	return err
+}
+
+func (fa *flowAggregator) InitAggregationProcess() error {
+	var err error
+	fa.aggregationProcess, err = ipfix.NewIPFIXAggregationProcess(fa.collectingProcess.GetMsgChan(), aggregationWorkerNum, aggregationFields)
+	return err
+}
+
+func (fa *flowAggregator) initExportingProcess() error {
+	obsID, err := genObservationID()
+	if err != nil {
+		return fmt.Errorf("cannot generate obsID for IPFIX ipfixexport: %v", err)
+	}
+	var ep ipfix.IPFIXExportingProcess
+	if fa.externalFlowCollectorAddr.Network() == "tcp" {
+		// TCP transport does not need any tempRefTimeout, so sending 0.
+		ep, err = ipfix.NewIPFIXExportingProcess(fa.externalFlowCollectorAddr, obsID, 0)
+	} else {
+		// For UDP transport, hardcoding tempRefTimeout value as 1800s.
+		ep, err = ipfix.NewIPFIXExportingProcess(fa.externalFlowCollectorAddr, obsID, 1800)
+	}
+	if err != nil {
+		return fmt.Errorf("got error when initializing IPFIX exporting process: %v", err)
+	}
+	fa.exportingProcess = ep
+	fa.templateID = fa.exportingProcess.NewTemplateID()
+	templateSet := ipfix.NewSet(ipfixentities.Template, fa.templateID, false)
+
+	bytesSent, err := fa.sendTemplateSet(templateSet)
+	if err != nil {
+		fa.exportingProcess.CloseConnToCollector()
+		fa.exportingProcess = nil
+		fa.templateID = 0
+		return fmt.Errorf("send template message failed, err: %v", err)
+	}
+	klog.V(2).Infof("Initialized exporting process and sent %d bytes size of template record", bytesSent)
+	return nil
 }
 
 func (fa *flowAggregator) Run(stopCh <-chan struct{}) {
-	// Place holder for Run() function, logic will be added in next PR
+	exportTicker := time.NewTicker(fa.exportInterval)
+	defer exportTicker.Stop()
+	go fa.collectingProcess.Start()
+	defer fa.collectingProcess.Stop()
+	go fa.aggregationProcess.Start()
+	defer fa.aggregationProcess.Stop()
+	for {
+		select {
+		case <-stopCh:
+			if fa.exportingProcess != nil {
+				fa.exportingProcess.CloseConnToCollector()
+			}
+			return
+		case <-exportTicker.C:
+			if fa.exportingProcess == nil {
+				err := fa.initExportingProcess()
+				if err != nil {
+					klog.Errorf("Error when initializing exporting process: %v", err)
+					// Initializing exporting process fails, will retry in next exportInterval
```

Comment:
maybe the error log should mention the retry: "Error when initializing exporting process, will retry in ..."

Reply:
Thanks, addressed.